### PR TITLE
Update Auction Module

### PR DESCRIPTION
**Pull request by rhuairahrighairidh**
Tests passing for auction

main TODOs:
- [ ] add much more detailed tests
  - [ ] auctions
  - [ ] maths
- [ ] add features
  - [ ] min bid increments
  - [ ] handle auctions expiring with no bids - probably restart
- [ ] handle chain restarts gracefully

other TODOs
- [ ] events
- [ ] custom error types
- [ ] user interfaces (cli, rest, querier)
- [ ] standardise String methods

### CODE REVIEW

**Comment by rhuairahrighairidh** [comment]
Agree these would be needed. Though there’s a couple of higher level decisions that are still hazy to me that would effect the requirements:
* Should unsold collateral be repaid to depositor’s accounts, or to the cdp deposits?
* Should auctions pay out continuously (ie for every bid, the increase over last bid is paid to auction initiator)?

Current code repays collateral to accounts, and pays out continuously.

Thoughts on structure in each case:
* If collateral repaid to accounts:
    * don’t need anything fancy - a new auction could be passed a list of deposits on start, then these are paid out to rateably from the auction module.
* If collateral repaid to cdp deposits, then liquidator/cdp needs to run logic for events happening in the auction keeper.
    * If auctions pay out at end, liquidator end blocker as suggested sounds easiest.
    * If they pay out continuously, sdk hooks model might be cleaner as they handle things occurring mid block.

File: x/auction/types/params.go [hunk 1]
```diff
@@ -3,96 +3,81 @@ package types
...
 const (
-	// DefaultMaxAuctionDuration max length of auction, roughly 2 days in blocks
-	DefaultMaxAuctionDuration EndTime = 2 * 24 * 3600 / 5
+	// DefaultMaxAuctionDuration max length of auction
+	DefaultMaxAuctionDuration time.Duration = 2 * 24 * time.Hour
 	// DefaultBidDuration how long an auction gets extended when someone bids, roughly 3 hours in blocks
-	DefaultMaxBidDuration EndTime = 3 * 3600 / 5
-	// DefaultStartingAuctionID what the id of the first auction will be
-	DefaultStartingAuctionID ID = ID(0)
+	DefaultBidDuration time.Duration = 3 * time.Hour
```
Comment (karzak) [inline x/auction/types/params.go]: I think this is too long. The main 'worry' with short bid durations is block stuffing to prevent other bids. The tradeoff with long timeouts is greater amount of locked capital required. I would start with something like 1 hour.

File: x/auction/types/params.go [hunk 2]
```diff
@@ -3,96 +3,81 @@ package types
...
-		return fmt.Errorf("max auction duration should be positive, is %s", ap.MaxAuctionDuration)
-	}
-	if ap.MaxBidDuration <= EndTime(0) {
-		return fmt.Errorf("bid duration should be positive, is %s", ap.MaxBidDuration)
-	}
-	if ap.StartingAuctionID <= ID(0) {
-		return fmt.Errorf("starting auction ID should be positive, is %v", ap.StartingAuctionID)
-	}
+func (ap Params) Validate() error {
+	// TODO check durations are within acceptable limits, if needed
```
Comment (karzak) [inline x/auction/types/params.go]: Implement this. Main checks would be that bid duration and auction duration are positive, maybe that they aren't too small.

File: x/auction/types/auctions.go [hunk 3]
```diff
@@ -111,117 +58,64 @@ func (a BaseAuction) String() string {
...
-		Lot:        lot,
-		Bidder:     seller,     // send the proceeds from the first bid back to the seller
-		Bid:        initialBid, // set this to zero most of the time
-		EndTime:    EndTime,
-		MaxEndTime: EndTime,
-	}
-	return auction
-}
-
 // ForwardAuction type for forward auctions
```
Comment (karzak) [inline x/auction/types/auctions.go]: After the go doc comment (next line so it doesn't get included in the api docs), I would include a quick note of when each auction is used.

File: x/auction/keeper/auctions.go [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+	err := k.supplyKeeper.SendCoinsFromModuleToModule(ctx, seller, types.ModuleName, sdk.NewCoins(lot))
+	if err != nil {
+		return 0, err
+	}
+	// store the auction
+	auctionID, err := k.StoreNewAuction(ctx, auction)
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: emit even for auction start

File: x/auction/keeper/auctions.go [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+	macc := k.supplyKeeper.GetModuleAccount(ctx, buyer)
+	if !macc.HasPermission(supply.Minter) { // TODO ideally don't want to import supply
+		return 0, sdk.ErrInternal("module does not have minting permissions")
+	}
+	// store the auction
+	auctionID, err := k.StoreNewAuction(ctx, auction)
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: same as above

File: x/auction/keeper/auctions.go [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+	err = k.supplyKeeper.SendCoinsFromModuleToModule(ctx, seller, types.ModuleName, sdk.NewCoins(lot))
+	if err != nil {
+		return 0, err
+	}
+	// store the auction
+	auctionID, err := k.StoreNewAuction(ctx, auction)
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: same as above

File: x/auction/keeper/auctions.go [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+import (
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/supply"
+	"github.com/kava-labs/kava/x/auction/types"
+)
+
+// StartForwardAuction starts a normal auction that mints the sold coins.
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: worth noting that all bids are burned for Forward Auctions - different than how someone might expect the auction to work.

File: x/auction/keeper/auctions.go [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+// StartReverseAuction starts an auction where sellers compete by offering decreasing prices.
+func (k Keeper) StartReverseAuction(ctx sdk.Context, buyer string, bid sdk.Coin, initialLot sdk.Coin) (uint64, sdk.Error) {
+	// create auction
+	auction := types.NewReverseAuction(buyer, bid, initialLot, ctx.BlockTime().Add(types.DefaultMaxAuctionDuration))
+
+	// This auction type mints coins at close. Need to check module account has minting privileges to avoid potential err in endblocker.
+	macc := k.supplyKeeper.GetModuleAccount(ctx, buyer)
+	if !macc.HasPermission(supply.Minter) { // TODO ideally don't want to import supply
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: remove todo, importing supply seems fine

File: x/auction/keeper/auctions.go [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+	}
+
+	// store updated auction
+	k.SetAuction(ctx, a)
+	return nil
+}
+
+func (k Keeper) PlaceBidForward(ctx sdk.Context, a types.ForwardAuction, bidder sdk.AccAddress, bid sdk.Coin) (types.ForwardAuction, sdk.Error) {
+	// Valid New Bid
+	if bid.Denom != a.Bid.Denom {
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: Isn't this validation performed already on line 84?

File: x/auction/keeper/auctions.go [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+		return a, sdk.ErrInternal("bid not greater than last bid")
+	}
+
+	// Move Coins
+	increment := bid.Sub(a.Bid)
+	bidAmtToReturn := a.Bid
+	if bidder.Equals(a.Bidder) { // catch edge case of someone updating their bid with a low balance
+		bidAmtToReturn = sdk.NewInt64Coin(a.Bid.Denom, 0)
+	}
+	err := k.supplyKeeper.SendCoinsFromAccountToModule(ctx, bidder, types.ModuleName, sdk.NewCoins(bidAmtToReturn.Add(increment)))
```
[the remark this answers is not in the document]
Comment (karzak) [reply x/auction/keeper/auctions.go]: Confirmed zero doesn't cause an error. I would still use swtich/ifelse to avoid these calls when coins is zero so we aren't emitting noisy events for transfering 0 coins.

[the remark this answers is not in the document]
Comment (karzak) [reply x/auction/keeper/auctions.go]: Also, could we just send `bidIncrement` straight to a.Initiator and avoid having to call `SendCoinsFromModuleToModule` on line 142?

File: x/auction/keeper/auctions.go [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+	// Validate New Bid // TODO min bid increments, make validation code less confusing
+	if !a.Bid.IsEqual(a.MaxBid) {
+		// Auction is in forward phase, a bid here can put the auction into forward or reverse phases
+		if !a.Bid.IsLT(bid) {
+			return a, sdk.ErrInternal("auction in forward phase, new bid not higher than last bid")
+		}
+		if a.MaxBid.IsLT(bid) {
+			return a, sdk.ErrInternal("bid higher than max bid")
+		}
+		if lot.IsNegative() || a.Lot.IsLT(lot) {
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: `lot.IsNegative() || a.Lot.IsLT(lot)` → `!lot.Equal(a.lot)`

File: x/auction/keeper/auctions.go [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, a.Bidder, sdk.NewCoins(bidAmtToReturn))
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, a.Initiator, sdk.NewCoins(bidIncrement))
+	if err != nil {
+		return a, err
+	}
+	// FIXME paying out rateably to cdp depositors is vulnerable to errors compounding over multiple bids
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: Given most coins are in 10^-6 units, rounding errors will generally be insignificant. Ok to note it, not sure it needs fixing.

File: x/auction/keeper/auctions.go [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+	a.EndTime = earliestTime(ctx.BlockTime().Add(types.DefaultBidDuration), a.MaxEndTime)
+
+	return a, nil
+}
+func (k Keeper) PlaceBidReverse(ctx sdk.Context, a types.ReverseAuction, bidder sdk.AccAddress, lot sdk.Coin) (types.ReverseAuction, sdk.Error) {
+	// Validate New Bid
+	if lot.Denom != a.Lot.Denom {
+		return a, sdk.ErrInternal("lot denom doesn't match auction")
+	}
+	if lot.IsNegative() {
```
Comment (karzak) [inline x/auction/keeper/auctions.go]: Is zero ok?

File: x/auction/abci.go [hunk 14]
```diff
@@ -7,17 +7,16 @@ import (
...
-	// get an iterator of expired auctions
-	expiredAuctions := k.GetQueueIterator(ctx, EndTime(ctx.BlockHeight()))
-	defer expiredAuctions.Close()
-
-	// loop through and close them - distribute funds, delete from store (and queue)
-	for ; expiredAuctions.Valid(); expiredAuctions.Next() {
-
-		auctionID := k.DecodeAuctionID(ctx, expiredAuctions.Value())
-		err := k.CloseAuction(ctx, auctionID)
+	var expiredAuctions []uint64
```
Comment (karzak) [inline x/auction/abci.go]: Move this logic to a keeper function called `CloseExpiredAuctions` and just call it here. Follows the cosmos-sdk pattern of not putting logic in `abci.go` and wrapping it in a keeper method.

File: x/auction/keeper/auctions_test.go [hunk 15]
```diff
@@ -0,0 +1,245 @@
+package keeper_test
```
Comment (karzak) [inline x/auction/keeper/auctions_test.go]: I'd like to see test coverage increase here. Main cases for each auction:
Same bidder, valid bid.
New bidder, valid bid
Add tests for main invalid bid cases

File: x/auction/keeper/auctions.go [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+	a.Bidder = bidder
+	a.Bid = bid
+	// increment timeout
+	a.EndTime = earliestTime(ctx.BlockTime().Add(types.DefaultBidDuration), a.MaxEndTime)
+
+	return a, nil
+}
+func (k Keeper) PlaceBidForwardReverse(ctx sdk.Context, a types.ForwardReverseAuction, bidder sdk.AccAddress, bid sdk.Coin, lot sdk.Coin) (types.ForwardReverseAuction, sdk.Error) {
+	// Validate New Bid // TODO min bid increments, make validation code less confusing
+	if !a.Bid.IsEqual(a.MaxBid) {
```
Comment (denalimarsh) [inline x/auction/keeper/auctions.go]: Replacing this check with `a.Phase == AuctionPhase.ForwardPhase` would make this more readable. Auction starts `ForwardPhase` and is updated to `ReversePhase` at the end of the function `if bid.IsEqual(a.MaxBid)`.

File: x/auction/keeper/auctions.go [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+		}
+		if lot.IsNegative() || a.Lot.IsLT(lot) {
+			return a, sdk.ErrInternal("lot out of bounds")
+		}
+		if lot.IsLT(a.Lot) && !bid.IsEqual(a.MaxBid) {
+			return a, sdk.ErrInternal("auction cannot enter reverse phase without bidding max bid")
+		}
+	} else {
+		// Auction is in reverse phase, it can never leave reverse phase
+		if !bid.IsEqual(a.MaxBid) {
```
Comment (denalimarsh) [inline x/auction/keeper/auctions.go]: Same as `!bid.Equal(a.bid)` once we're in reverse phase

File: x/auction/keeper/auctions.go [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+	}
+	err := k.supplyKeeper.SendCoinsFromAccountToModule(ctx, bidder, types.ModuleName, sdk.NewCoins(bidAmtToReturn.Add(bidIncrement)))
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, a.Bidder, sdk.NewCoins(bidAmtToReturn))
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, a.Initiator, sdk.NewCoins(bidIncrement))
```
Comment (denalimarsh) [inline x/auction/keeper/auctions.go]: `bidIncrement` is always 0 in reverse phase, resulting in a transfer of 0 coins. Could place inside a reverse vs. forward phase conditional to prevent this transfer (see below).

File: x/auction/keeper/auctions.go [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+		}
+	default:
+		panic("unrecognized auction type")
+	}
+
+	// Delete auction from store (and queue)
+	k.DeleteAuction(ctx, auctionID)
+	return nil
+}
+func (k Keeper) MintAndPayoutAuctionLot(ctx sdk.Context, a types.ReverseAuction) sdk.Error {
```
Comment (denalimarsh) [inline x/auction/keeper/auctions.go]: add newline

File: x/auction/keeper/auctions.go [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+	if err != nil {
+		return err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, a.Initiator, a.Bidder, sdk.NewCoins(a.Lot))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func (k Keeper) PayoutAuctionLot(ctx sdk.Context, a types.Auction) sdk.Error {
```
Comment (denalimarsh) [inline x/auction/keeper/auctions.go]: add newline

File: x/auction/keeper/auctions.go [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+	case types.ReverseAuction:
+		err = k.MintAndPayoutAuctionLot(ctx, auc)
+		if err != nil {
+			return err
+		}
+	default:
+		panic("unrecognized auction type")
+	}
+
+	// Delete auction from store (and queue)
```
Comment (denalimarsh) [inline x/auction/keeper/auctions.go]: `(and queue)` -> `and byTime index`

File: x/auction/keeper/keeper.go [hunk 22]
```diff
@@ -1,314 +1,169 @@
...
-
-func (k Keeper) getNextAuctionIDKey() []byte {
-	return []byte("nextAuctionID")
-}
-func (k Keeper) getAuctionKey(auctionID types.ID) []byte {
-	return []byte(fmt.Sprintf("auctions:%d", auctionID))
+// InsertIntoIndex adds an auction ID and end time into the byTime index
+func (k Keeper) InsertIntoIndex(ctx sdk.Context, endTime time.Time, auctionID uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.AuctionByTimeKeyPrefix)
+	store.Set(types.GetAuctionByTimeKey(endTime, auctionID), types.Uint64ToBytes(auctionID)) // TODO
```
Comment (denalimarsh) [inline x/auction/keeper/keeper.go]: Leftover TODO? If intentional, add comment.

File: x/auction/spec/06_begin_block.md [hunk 23]
```diff
@@ -0,0 +1,20 @@
+<!--
+order: 6
+-->
+
+At the beginning of each block, auctions that have reached `EndTime` are closed. The logic to close auctions is as follows:
```
Comment (denalimarsh) [inline x/auction/spec/06_begin_block.md]: Spec states auctions are closed at the beginning of each block, but `abci.go` defines `EndBlocker`.

File: x/auction/types/auctions.go [hunk 24]
```diff
@@ -231,77 +125,47 @@ func (a ForwardReverseAuction) String() string {
   End Time:   						%s
 	Max End Time:      			%s
 	Max Bid									%s
-	Other Person						%s`,
+	LotReturns						%s`,
```
Comment (denalimarsh) [inline x/auction/types/auctions.go]: minor formatting - could we standardize the spacing on these `String` methods?

File: x/auction/types/expected_keepers.go [hunk 25]
```diff
@@ -2,9 +2,20 @@ package types
...
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	supplyexported "github.com/cosmos/cosmos-sdk/x/supply/exported"
 )
 
-type BankKeeper interface {
-	SubtractCoins(sdk.Context, sdk.AccAddress, sdk.Coins) (sdk.Coins, sdk.Error)
-	AddCoins(sdk.Context, sdk.AccAddress, sdk.Coins) (sdk.Coins, sdk.Error)
+// SupplyKeeper defines the expected supply Keeper
+type SupplyKeeper interface {
+	//GetSupply(ctx sdk.Context) supplyexported.SupplyI
```
Comment (denalimarsh) [inline x/auction/types/expected_keepers.go]: Remove comment

File: x/auction/types/expected_keepers.go [hunk 26]
```diff
@@ -2,9 +2,20 @@ package types
...
+
+	//GetModuleAddress(name string) sdk.AccAddress
```
Comment (denalimarsh) [inline x/auction/types/expected_keepers.go]: Same as above

File: x/auction/types/genesis.go [hunk 27]
```diff
@@ -4,26 +4,28 @@ import (
...
+// Auctions type for an array of auctions
+type Auctions []Auction
 
 // GenesisState - auction state that must be provided at genesis
 type GenesisState struct {
-	AuctionParams AuctionParams   `json:"auction_params" yaml:"auction_params"`
-	Auctions      GenesisAuctions `json:"genesis_auctions" yaml:"genesis_auctions"`
+	NextAuctionID uint64          `json:"next_auction_id" yaml:"next_auction_id"`
+	Params        Params          `json:"auction_params" yaml:"auction_params"`
+	Auctions      Auctions `json:"genesis_auctions" yaml:"genesis_auctions"`
```
Comment (denalimarsh) [inline x/auction/types/genesis.go]: minor formatting - add spacing between `Auctions` and `json`

File: x/auction/keeper/auctions.go [hunk 28]
```diff
@@ -0,0 +1,337 @@
...
+	k.DeleteAuction(ctx, auctionID)
+	return nil
+}
+func (k Keeper) MintAndPayoutAuctionLot(ctx sdk.Context, a types.ReverseAuction) sdk.Error {
+	err := k.supplyKeeper.MintCoins(ctx, a.Initiator, sdk.NewCoins(a.Lot))
+	if err != nil {
+		return err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, a.Initiator, a.Bidder, sdk.NewCoins(a.Lot))
```
Comment (denalimarsh) [inline x/auction/keeper/auctions.go]: Should this be `a.GetBidder()`, `a.GetLot()`? It looks like auction's getter methods aren't fully leveraged in other functions as well.

File: x/auction/types/expected_keepers.go (hunk 25 above)
Comment (denalimarsh) [reply x/auction/types/expected_keepers.go]: Duplicate of https://github.com/Kava-Labs/kava/pull/276#discussion_r364234919

**Comment by denalimarsh** [comment]
@rhuairahrighairidh please note that comments with single line selections in my most recent review seem to be displayed as 4 line blocks, see https://github.com/Kava-Labs/kava/pull/275#discussion_r364276095.

File: x/auction/keeper/auctions.go (hunk 28 above)
Comment (rhuairahrighairidh) [reply x/auction/keeper/auctions.go]: I switched things about to remove this. My intention was to only have getter and setters for code that deals with all auction types.

File: x/auction/keeper/auctions.go (hunk 16 above)
Comment (rhuairahrighairidh) [reply x/auction/keeper/auctions.go]: I've added a auction method `IsReversePhase` to make things clearer.
Thought this might be better than an enum on the struct as the phase can be determined from the fields that are already present.

File: x/auction/keeper/auctions.go [hunk 29]
```diff
@@ -0,0 +1,337 @@
...
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, a.Bidder, sdk.NewCoins(bidAmtToReturn))
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, a.Initiator, sdk.NewCoins(bidIncrement))
+	if err != nil {
+		return a, err
+	}
+	// FIXME paying out rateably to cdp depositors is vulnerable to errors compounding over multiple bids
+	lotPayouts, err := splitCoinIntoWeightedBuckets(lotDecrement, a.LotReturns.Weights)
```
[the remark this answers is not in the document]
Comment (rhuairahrighairidh) [reply x/auction/keeper/auctions.go]: I've removed all sending of zero coins by a combination of:
- adding conditionals for if the amounts are zero
- splitting the forward reverse function up into two functions - one for each phase

This does mean in order to switch between phases a bid must be placed with exactly the maxBid. But I think client code can make the UX better by creating a tx with two msgs for any bid between phases.

File: x/auction/keeper/auctions.go (hunk 7 above)
Comment (rhuairahrighairidh) [reply x/auction/keeper/auctions.go]: added more comments onto the auction types

File: x/auction/abci.go (hunk 14 above)
Comment (rhuairahrighairidh) [reply x/auction/abci.go]: good idea

**Comment by rhuairahrighairidh** [comment]
Still some TODOs, I've updated the top description with them, but I think we can merge and do them as individual PRs.

Major changes since last review:
- Auctions renamed, since they weren't really general purpose forward / reverse auctions.
- I split placing a bid on a FR ("collateral auction") into placing forward and reverse bids. This makes the logic a bit easier to follow I think, and MsgPlaceBid smaller.
- Cases where it would send zero coins have been removed

File: x/auction/keeper/auctions.go (hunk 28 above)
Comment (denalimarsh) [reply x/auction/keeper/auctions.go]: Okay, that makes sense.

File: x/auction/keeper/auctions.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/auction/keeper/auctions.go]: I think that making auction phase conditional upon a max bid makes sense (certainly for now).

**Comment by denalimarsh** [comment]
Could we add a few more tests? It's super close but think it's worth aiming for a minimum benchmark of ~80% coverage for the keeper directory and ~40% for both the types/module directory to keep our coverage comparable with Cosmos modules.

auction_module_test_coverage

Other than that, happy to approve this PR.
